src/relax/analysis/liveness.cc

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file tvm/relax/analysis/liveness.cc
+ * \brief Implementation of liveness analysis
+ */
+#include <tvm/relax/analysis.h>
+#include <tvm/relax/dataflow_analysis.h>
+#include <tvm/relax/expr.h>
+#include <tvm/relax/expr_functor.h>
+#include <tvm/runtime/object.h>
+
+namespace tvm {
+namespace relax {
+
+// just sets of vars. the bool value is unnecessary
+using Domain = Map<Var, Bool>;
+
+Domain transfer_func(const GraphBinding& binding, const ObjectRef& input) {
+  Domain in_domain = Downcast<Domain>(input);
+  Domain new_domain(in_domain);
+
+  // 1. If a var that appears in the RHS of the binding, add it (it's live)
+  // 2. Remove the bound var (it is not live prior to being bound)
+  Array<Var> vars_used;
+  Optional<Var> var_bound;
+  if (binding->kind == BindingNodeKind::kSeqBody) {
+    vars_used = AllVars(binding->seq->body);
+  } else if (binding->kind == BindingNodeKind::kIfCond) {
+    Binding b = binding->seq->blocks[binding->block_idx]->bindings[binding->binding_idx];
+    Expr cond = Downcast<If>(GetBoundValue(b))->cond;
+    vars_used = AllVars(cond);
+  } else if (binding->kind == BindingNodeKind::kIfMerge) {
+    // no vars are used in the merge
+    vars_used = {};
+    // define the merge var
+    var_bound = binding->seq->blocks[binding->block_idx]->bindings[binding->binding_idx]->var;
+  } else {
+    // the ordinary binding case
+    Binding b = binding->seq->blocks[binding->block_idx]->bindings[binding->binding_idx];
+    Expr bound_value = GetBoundValue(b);
+    // special case: if the RHS is a function literal, we only care about the free vars
+    // (those captured by the closure)
+    if (bound_value.as<FunctionNode>()) {
+      vars_used = FreeVars(bound_value);
+    } else {
+      vars_used = AllVars(bound_value);
+    }
+    var_bound = b->var;
+  }
+
+  for (auto var : vars_used) {
+    if (!new_domain.count(var)) {
```

Comment:
Nit: Both `.count` and `.Set` need to hash the key and find the index in the map (or if it gets changed to `unordered_set`, the set).  Since the contents are already being de-duplicated, we could just call `new_domain.Set(var, Bool(true))` without the `if(!new_domain.count(var)`.  If `var` is already present, it's a no-op, and if it isn't already present, we've saved a lookup.

Reply:
I wasn't sure if it worked that way, so that's good. I'll take note of that, thanks.